### ISSUE TITLE
Player Emotes, Animations, and Leantween Networking Bounty B:100

### ISSUE THREAD

**Issue by PerfectTangent** [unitystation/unitystation#6157]
## Description

This bounty calls for the addition of all the most common player emotes, including the ones that have accompanying animations. It also requires animations for when players pick up items and for when players for down as depicted here #6074. 

_Furthermore this bounty should include a generic or standard way to network leantween animations so it is easier for contributors to implement new animations that all clients should be able to see_

## What is needed to fulfill the bounty:

- [x] add all of /tg/'s player emotes.
- [x] add all appropriate character animations for said emotes, where applicable
- [x] add an animation for picking up items.
- [x] add an animation for characters falling to the ground.
- [x] add a generic way to network leantween animation.

**Comment by MaxIsJoe** [comment]
I'm up for the task. Going to work on this.

**Comment by MaxIsJoe** [comment]
First four hours while working on this and I have already made progress!
![image](https://user-images.githubusercontent.com/34368774/109607622-2dfb7400-7b31-11eb-8709-8e723d703994.png)

Created a modular system that can detect specific emotes and run special behaviors associated with those emotes. Just pop an EmoteAction script to the emote manager and boom, any code under that script will run if the game finds the right emote name.
![image](https://user-images.githubusercontent.com/34368774/109607852-87fc3980-7b31-11eb-8aca-3737a31957e1.png)

Next up :
- Making the player's name display when doing a special emote
- Add in all emotes in TG
- start working on animations after all that is done.

**Comment by corp-0** [comment]
Looking good 👌

**Comment by corp-0** [comment]
where are the components added?

**Comment by NoooneyDude** [comment]
I'm a little lost as to why `Backflip` is its own component. You'd have one component for all emotes, right? And it'd have the emote name, text fields and, perhaps, an SO for the actual animation assigned. Then you can just tell the clients to animate via the SO that they all have.

How you define the animation SO is the key to networking it to clients and making it reusible for other systems that require tweening.

In fact, perhaps the emotes would be better defined in an SO themselves, too, though I suppose you have something worked out with having them as component

**Comment by MaxIsJoe** [comment]
> I'm a little lost as to why `Backflip` is its own component. You'd have one component for all emotes, right? And it'd have the emote name, text fields and, perhaps, an SO for the actual animation assigned. Then you can just tell the clients to animate via the SO that they all have.
> 
> How you define the animation SO is the key to networking it to clients and making it reusible for other systems that require tweening.
> 
> In fact, perhaps the emotes would be better defined in an SO themselves, too, though I suppose you have something worked out with having them as component

The reason why I made it this way is because it keeps things simple and more flexible to work with, Animations will use LeanTween by default with my additions that will (hopefully) try making animation syncing much easier and cleaner.
All emotes currently inherit from an Emote class which will hopefully provide all the necessary things that won't make people rewrite code over and over again for different cases.
![image](https://user-images.githubusercontent.com/34368774/109609058-7a47b380-7b33-11eb-9139-2ec9dde0d4f4.png)

P.S : I did try approaching this using scriptable objects, was not very pleasant to work with.

**Comment by NoooneyDude** [comment]
Okay, cool. So long as you have a plan for creating a system to network the tweening (that can be used by other systems), which is by far the most difficult task here.

**Comment by corp-0** [comment]
why was using SO's not pleasant ?

**Comment by MaxIsJoe** [comment]
> why was using SO's not pleasant ?

It's mostly regarding organizing stuff and going over the same question of "how do I call or do this from an SO without hardcoding a solution?"
however, if you guys want me to make this use SOs then now's the earliest and best time to make the change.. though that means I'd probably have to rework a lot of the emotes that I did not commit yet lol.

**Comment by corp-0** [comment]
we can discuss it on discord to find the best approach. Using SO's shouldn't be unpleseant :B We use it for lots of stuff without hardcoding behaviour.

**Comment by MaxIsJoe** [comment]
> we can discuss it on discord to find the best approach. Using SO's shouldn't be unpleseant :B We use it for lots of stuff without hardcoding behaviour.

Alrighty, just send me a DM at `/dev/Max#6049` and we can talk about how I should approach this.

**Comment by MaxIsJoe** [comment]
Some updates.
SOs are now used for emotes (if you didn't see the last two commits above this) and have now added some extra functionality to make people write cleaner code when creating complex emotes in the future.

**Player checks**
![image](https://user-images.githubusercontent.com/34368774/110248368-5ba14c80-7f79-11eb-8be0-d8fe86450c11.png)
![image](https://user-images.githubusercontent.com/34368774/110248210-9ce52c80-7f78-11eb-9873-3d8d670bc053.png)

For emotes that require the player's gender to be checked, I have created a new SO that inherits from the base emote class that checks the player's gender.
It also takes account the state that the player's health is in to show some special text for others around them.

**Audio**
Audio uses addressables by default but I have not added any audio files for emotes *yet* because I don't think it's worth my time to setup now. maybe later when I'm done with creating the last special emotes that require behaviors and the rest of networkedLeanTween.

overall I'd say that 95% of all /TG/ emotes have been added. maybe three special ones left and that's it.
